File: python/notifier.py
"""Email notifications for bot state changes.

Configure via env vars:
    EMAIL_ENABLED=true
    EMAIL_SMTP_HOST=smtp.gmail.com
    EMAIL_SMTP_PORT=587
    EMAIL_USE_TLS=true         (STARTTLS; set false for SMTP_SSL on port 465)
    EMAIL_USER=mybot@gmail.com
    EMAIL_PASSWORD=app_password
    EMAIL_TO=me@example.com
"""

import logging
import smtplib
import ssl
from datetime import datetime
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

log = logging.getLogger("bot.notifier")


class Notifier:
    """Sends email notifications for bot state changes.

    All send methods silently swallow errors — a notification failure must
    never crash the bot.
    """

    def __init__(self, config):
        self._config = config

    @property
    def enabled(self) -> bool:
        return (
            self._config.email_enabled
            and bool(self._config.email_smtp_host)
            and bool(self._config.email_to)
        )

    def send(self, subject: str, body: str) -> None:
        """Send a plain-text email. No-op if notifications are disabled."""
        if not self.enabled:
            return
        cfg = self._config
        try:
            msg = MIMEMultipart("alternative")
            msg["Subject"] = f"[Polymarket Bot] {subject}"
            msg["From"] = cfg.email_user or f"polymarket-bot@{cfg.email_smtp_host}"
            msg["To"] = cfg.email_to
            msg.attach(MIMEText(body, "plain"))

            context = ssl.create_default_context()
            if cfg.email_use_tls:
                # STARTTLS (port 587)
                with smtplib.SMTP(cfg.email_smtp_host, cfg.email_smtp_port) as smtp:
                    smtp.ehlo()
                    smtp.starttls(context=context)
                    smtp.ehlo()
                    if cfg.email_user and cfg.email_password:
                        smtp.login(cfg.email_user, cfg.email_password)
                    smtp.sendmail(msg["From"], [cfg.email_to], msg.as_string())
            else:
                # SMTP_SSL (port 465)
                with smtplib.SMTP_SSL(cfg.email_smtp_host, cfg.email_smtp_port, context=context) as smtp:
                    if cfg.email_user and cfg.email_password:
                        smtp.login(cfg.email_user, cfg.email_password)
                    smtp.sendmail(msg["From"], [cfg.email_to], msg.as_string())

            log.debug(f"Email sent: {subject}")
        except Exception as e:
            log.warning(f"Email notification failed: {e}")

    # ── Convenience helpers ────────────────────────────────────────────────

    def notify_started(self, mode: str, bankroll: float, positions: int) -> None:
        self.send(
            f"Started: {mode} mode",
            f"Bot started.\n\n"
            f"Mode: {mode}\n"
            f"Bankroll: ${bankroll:.2f}\n"
            f"Open positions: {positions}\n"
            f"Time: {_now()}",
        )

    def notify_trade(self, trade, signal, portfolio) -> None:
        self.send(
            f"BUY {trade.side.value} ${trade.size_usd:.2f} — {trade.question[:60]}",
            f"New position opened.\n\n"
            f"Market: {trade.question}\n"
            f"Side: {trade.side.value}\n"
            f"Price: {trade.price:.4f}\n"
            f"Size: ${trade.size_usd:.2f}\n"
            f"Shares: {trade.shares:.2f}\n"
            f"Edge: {signal.edge:.1%}\n"
            f"Expected value: ${signal.expected_value:.2f}\n"
            f"\nPortfolio after:\n"
            f"  Bankroll: ${portfolio.bankroll:.2f}\n"
            f"  Exposure: ${portfolio.total_exposure():.2f}\n"
            f"  Positions: {len(portfolio.positions)}\n"
            f"Time: {_now()}",
        )

    def notify_sell(self, trade, exit_reason: str, pnl_pct: float, portfolio) -> None:
        sign = "+" if pnl_pct >= 0 else ""
        self.send(
            f"SELL ({exit_reason}) {sign}{pnl_pct:.1%} — {trade.question[:60]}",
            f"Position closed.\n\n"
            f"Market: {trade.question}\n"
            f"Exit reason: {exit_reason}\n"
            f"Exit price: {trade.price:.4f}\n"
            f"PnL: {sign}{pnl_pct:.1%}\n"
            f"Recovered: ${trade.size_usd:.2f}\n"
            f"\nPortfolio after:\n"
            f"  Bankroll: ${portfolio.bankroll:.2f}\n"
            f"  Exposure: ${portfolio.total_exposure():.2f}\n"
            f"  Positions: {len(portfolio.positions)}\n"
            f"  Realized PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )

    def notify_topup_sell(self, trade, tc, portfolio) -> None:
        self.send(
            f"TOPUP+SELL ({tc.exit_reason}) recovered ${tc.recovery_value:.2f} — {tc.position.question[:55]}",
            f"Tiny position rescued via top-up-and-sell.\n\n"
            f"Market: {tc.position.question}\n"
            f"Exit reason: {tc.exit_reason}\n"
            f"Tokens bought: {tc.tokens_to_buy:.0f} (top-up)\n"
            f"Total tokens sold: {tc.position.shares + tc.tokens_to_buy:.2f}\n"
            f"Top-up cost: ${tc.topup_cost:.2f}\n"
            f"Recovered: ${tc.recovery_value:.2f}\n"
            f"\nPortfolio after:\n"
            f"  Bankroll: ${portfolio.bankroll:.2f}\n"
            f"  Exposure: ${portfolio.total_exposure():.2f}\n"
            f"  Positions: {len(portfolio.positions)}\n"
            f"  Realized PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )

    def notify_resolved(self, position, won: bool, pnl: float, portfolio) -> None:
        result = "WON" if won else "LOST"
        self.send(
            f"Resolved ({result}) PnL={pnl:+.2f} — {position.question[:60]}",
            f"Market resolved.\n\n"
            f"Market: {position.question}\n"
            f"Result: {result}\n"
            f"PnL: ${pnl:+.2f}\n"
            f"Shares: {position.shares:.2f}\n"
            f"\nPortfolio after:\n"
            f"  Bankroll: ${portfolio.bankroll:.2f}\n"
            f"  Exposure: ${portfolio.total_exposure():.2f}\n"
            f"  Positions: {len(portfolio.positions)}\n"
            f"  Realized PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )

    def notify_halted(self, reason: str, portfolio) -> None:
        pv = portfolio.bankroll + portfolio.total_exposure()
        self.send(
            f"HALTED: {reason}",
            f"Bot halted.\n\n"
            f"Reason: {reason}\n"
            f"\nPortfolio:\n"
            f"  Value: ${pv:.2f}\n"
            f"  Bankroll: ${portfolio.bankroll:.2f}\n"
            f"  Exposure: ${portfolio.total_exposure():.2f}\n"
            f"  Positions: {len(portfolio.positions)}\n"
            f"  Realized PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )

    def notify_daily_reset(self, portfolio) -> None:
        pv = portfolio.bankroll + portfolio.total_exposure()
        self.send(
            f"Daily reset — portfolio ${pv:.2f}",
            f"New trading day started.\n\n"
            f"Portfolio value: ${pv:.2f}\n"
            f"Bankroll: ${portfolio.bankroll:.2f}\n"
            f"Exposure: ${portfolio.total_exposure():.2f}\n"
            f"Open positions: {len(portfolio.positions)}\n"
            f"Cumulative PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )

    def notify_error(self, cycle: int, error: Exception) -> None:
        self.send(
            f"Error in cycle {cycle}",
            f"An error occurred in cycle {cycle}.\n\n"
            f"Error: {error}\n"
            f"Time: {_now()}",
        )

    def notify_stopped(self, portfolio) -> None:
        pv = portfolio.bankroll + portfolio.total_exposure()
        self.send(
            f"Stopped — portfolio ${pv:.2f}, PnL {portfolio.total_realized_pnl:+.2f}",
            f"Bot stopped.\n\n"
            f"Final portfolio value: ${pv:.2f}\n"
            f"Bankroll: ${portfolio.bankroll:.2f}\n"
            f"Exposure: ${portfolio.total_exposure():.2f}\n"
            f"Open positions: {len(portfolio.positions)}\n"
            f"Total trades: {portfolio.total_trades}\n"
            f"Total API cost: ${portfolio.total_api_cost:.4f}\n"
            f"Realized PnL: ${portfolio.total_realized_pnl:+.2f}\n"
            f"Time: {_now()}",
        )


def _now() -> str:
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

File: dotnet/PolymarketBot/Services/Notifier.cs
using System.Net;
using System.Net.Mail;
using Microsoft.Extensions.Logging;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

/// <summary>
/// Sends email notifications for bot state changes.
/// All methods silently swallow errors — a notification failure must never crash the bot.
/// Configure via env vars: EMAIL_ENABLED, EMAIL_SMTP_HOST, EMAIL_SMTP_PORT,
/// EMAIL_USE_TLS, EMAIL_USER, EMAIL_PASSWORD, EMAIL_TO.
/// </summary>
public sealed class Notifier
{
    private readonly BotConfig _config;
    private readonly ILogger<Notifier> _logger;

    public Notifier(BotConfig config, ILogger<Notifier> logger)
    {
        _config = config;
        _logger = logger;
    }

    public bool Enabled =>
        _config.EmailEnabled &&
        !string.IsNullOrEmpty(_config.EmailSmtpHost) &&
        !string.IsNullOrEmpty(_config.EmailTo);

    public void Send(string subject, string body)
    {
        if (!Enabled) return;
        try
        {
#pragma warning disable SYSLIB0021 // SmtpClient is deprecated but acceptable for simple SMTP
            using var client = new SmtpClient(_config.EmailSmtpHost, _config.EmailSmtpPort)
            {
                EnableSsl = _config.EmailUseTls,
                DeliveryMethod = SmtpDeliveryMethod.Network,
                UseDefaultCredentials = false,
            };
#pragma warning restore SYSLIB0021
            if (!string.IsNullOrEmpty(_config.EmailUser))
                client.Credentials = new NetworkCredential(_config.EmailUser, _config.EmailPassword);

            var from = string.IsNullOrEmpty(_config.EmailUser)
                ? $"polymarket-bot@{_config.EmailSmtpHost}"
                : _config.EmailUser;

            using var msg = new MailMessage(from, _config.EmailTo)
            {
                Subject = $"[Polymarket Bot] {subject}",
                Body = body,
                IsBodyHtml = false,
            };
            client.Send(msg);
            _logger.LogDebug("Email sent: {Subject}", subject);
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Email notification failed: {Error}", ex.Message);
        }
    }

    // ── Convenience helpers ──────────────────────────────────────────────

    public void NotifyStarted(string mode, double bankroll, int positions) =>
        Send($"Started: {mode} mode",
            $"Bot started.\n\n" +
            $"Mode: {mode}\n" +
            $"Bankroll: ${bankroll:F2}\n" +
            $"Open positions: {positions}\n" +
            $"Time: {Now()}");

    public void NotifyTrade(Trade trade, Signal signal, Portfolio portfolio) =>
        Send($"BUY {trade.Side} ${trade.SizeUsd:F2} — {Truncate(trade.Question, 60)}",
            $"New position opened.\n\n" +
            $"Market: {trade.Question}\n" +
            $"Side: {trade.Side}\n" +
            $"Price: {trade.Price:F4}\n" +
            $"Size: ${trade.SizeUsd:F2}\n" +
            $"Shares: {trade.Shares:F2}\n" +
            $"Edge: {signal.Edge:P1}\n" +
            $"Expected value: ${signal.ExpectedValue:F2}\n" +
            $"\nPortfolio after:\n" +
            $"  Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"  Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"  Positions: {portfolio.Positions.Count}\n" +
            $"Time: {Now()}");

    public void NotifySell(Trade trade, string exitReason, double pnlPct, Portfolio portfolio)
    {
        var sign = pnlPct >= 0 ? "+" : "";
        Send($"SELL ({exitReason}) {sign}{pnlPct:P1} — {Truncate(trade.Question, 60)}",
            $"Position closed.\n\n" +
            $"Market: {trade.Question}\n" +
            $"Exit reason: {exitReason}\n" +
            $"Exit price: {trade.Price:F4}\n" +
            $"PnL: {sign}{pnlPct:P1}\n" +
            $"Recovered: ${trade.SizeUsd:F2}\n" +
            $"\nPortfolio after:\n" +
            $"  Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"  Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"  Positions: {portfolio.Positions.Count}\n" +
            $"  Realized PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");
    }

    public void NotifyTopupSell(Trade trade, TopupCandidate tc, Portfolio portfolio) =>
        Send($"TOPUP+SELL ({tc.ExitReason}) recovered ${tc.RecoveryValue:F2} — {Truncate(tc.Position.Question, 55)}",
            $"Tiny position rescued via top-up-and-sell.\n\n" +
            $"Market: {tc.Position.Question}\n" +
            $"Exit reason: {tc.ExitReason}\n" +
            $"Tokens bought: {tc.TokensToBuy:F0} (top-up)\n" +
            $"Total tokens sold: {tc.Position.Shares + tc.TokensToBuy:F2}\n" +
            $"Top-up cost: ${tc.TopupCost:F2}\n" +
            $"Recovered: ${tc.RecoveryValue:F2}\n" +
            $"\nPortfolio after:\n" +
            $"  Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"  Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"  Positions: {portfolio.Positions.Count}\n" +
            $"  Realized PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");

    public void NotifyResolved(Position pos, bool won, double pnl, Portfolio portfolio) =>
        Send($"Resolved ({(won ? "WON" : "LOST")}) PnL={pnl:+0.00;-0.00} — {Truncate(pos.Question, 60)}",
            $"Market resolved.\n\n" +
            $"Market: {pos.Question}\n" +
            $"Result: {(won ? "WON" : "LOST")}\n" +
            $"PnL: ${pnl:+0.00;-0.00}\n" +
            $"Shares: {pos.Shares:F2}\n" +
            $"\nPortfolio after:\n" +
            $"  Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"  Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"  Positions: {portfolio.Positions.Count}\n" +
            $"  Realized PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");

    public void NotifyHalted(string reason, Portfolio portfolio)
    {
        var pv = portfolio.Bankroll + portfolio.TotalExposure();
        Send($"HALTED: {reason}",
            $"Bot halted.\n\n" +
            $"Reason: {reason}\n" +
            $"\nPortfolio:\n" +
            $"  Value: ${pv:F2}\n" +
            $"  Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"  Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"  Positions: {portfolio.Positions.Count}\n" +
            $"  Realized PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");
    }

    public void NotifyDailyReset(Portfolio portfolio)
    {
        var pv = portfolio.Bankroll + portfolio.TotalExposure();
        Send($"Daily reset — portfolio ${pv:F2}",
            $"New trading day started.\n\n" +
            $"Portfolio value: ${pv:F2}\n" +
            $"Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"Open positions: {portfolio.Positions.Count}\n" +
            $"Cumulative PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");
    }

    public void NotifyBuyFail(MarketInfo market, Signal signal, string reason) =>
        Send($"BUY FAILED {signal.Side} ${signal.PositionSizeUsd:F2} — {Truncate(market.Question, 60)}",
            $"BUY order failed.\n\n" +
            $"Market: {market.Question}\n" +
            $"Side: {signal.Side}\n" +
            $"Attempted price: {signal.MarketPrice:F4}\n" +
            $"Attempted size: ${signal.PositionSizeUsd:F2}\n" +
            $"Edge: {signal.Edge:P1}\n" +
            $"Reason: {reason}\n" +
            $"Time: {Now()}");

    public void NotifySellFail(Position position, string exitReason, string failReason) =>
        Send($"SELL FAILED ({exitReason}) — {Truncate(position.Question, 60)}",
            $"SELL order failed.\n\n" +
            $"Market: {position.Question}\n" +
            $"Exit reason: {exitReason}\n" +
            $"Attempted price: {position.CurrentPrice:F4}\n" +
            $"Shares: {position.Shares:F2}\n" +
            $"Reason: {failReason}\n" +
            $"Time: {Now()}");

    public void NotifyTopupSellFail(TopupCandidate tc, string failReason) =>
        Send($"TOPUP+SELL FAILED ({tc.ExitReason}) — {Truncate(tc.Position.Question, 55)}",
            $"Top-up-and-sell failed.\n\n" +
            $"Market: {tc.Position.Question}\n" +
            $"Exit reason: {tc.ExitReason}\n" +
            $"Current tokens: {tc.Position.Shares:F2}\n" +
            $"Top-up cost: ${tc.TopupCost:F2}\n" +
            $"Reason: {failReason}\n" +
            $"Time: {Now()}");

    public void NotifyError(int cycle, Exception ex) =>
        Send($"Error in cycle {cycle}",
            $"An error occurred in cycle {cycle}.\n\n" +
            $"Error: {ex.Message}\n" +
            $"Time: {Now()}");

    public void NotifyStopped(Portfolio portfolio)
    {
        var pv = portfolio.Bankroll + portfolio.TotalExposure();
        Send($"Stopped — portfolio ${pv:F2}, PnL {portfolio.TotalRealizedPnl:+0.00;-0.00}",
            $"Bot stopped.\n\n" +
            $"Final portfolio value: ${pv:F2}\n" +
            $"Bankroll: ${portfolio.Bankroll:F2}\n" +
            $"Exposure: ${portfolio.TotalExposure():F2}\n" +
            $"Open positions: {portfolio.Positions.Count}\n" +
            $"Total trades: {portfolio.TotalTrades}\n" +
            $"Total API cost: ${portfolio.TotalApiCost:F4}\n" +
            $"Realized PnL: ${portfolio.TotalRealizedPnl:+0.00;-0.00}\n" +
            $"Time: {Now()}");
    }

    private static string Now() => DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
    private static string Truncate(string s, int maxLen) => s.Length <= maxLen ? s : s[..maxLen] + "...";
}
